src/build.rs

```diff
@@ -263,3 +265,323 @@ fn build_nasm(root: &Path, vendor_dir: &Path, out_dir: &Path, target_arch: &str,
+            params = decl.args.iter().map(|(id, ty)| c_param(&id, &ty)).collect::<Vec<_>>().join(","),
+            value_return = value_return
+    )
+}
+
+fn is_c_ptr(v: &str) -> bool { v.starts_with("&") || v.starts_with("*") }
+
+fn rust_ptr_to_c_ptr(x: &str) -> String { // ADAPTED FROM CTEST
+    let mut ty = String::new();
+    for x in x.split('\'') { 
+        if ty.is_empty() {
+            ty += x;
+        } else {
+            if let Some(i) = x.find(char::is_whitespace) {
+                ty += &x[i..];
+            } else {
+                break;
+            }
+        }
+    }
+    let mut mutability = Vec::new();
+    let ptr_count = ty.chars().filter(|c| is_c_ptr(&c.to_string())).count();
+    let mut ty: &str = &ty[1..];
+    for _ in 0..ptr_count {
+        if let Some(idx) = ty.find(|c: char| is_c_ptr(&c.to_string())) {
+            if let Some(mut_idx) = ty[0..idx].find("mut") {
+                mutability.push(true);
+                ty = &ty[mut_idx+5..];
+            } else {
+                mutability.push(false);
+                if let Some(const_idx) = ty[0..idx].find("const") {
+                    ty = &ty[const_idx+6..];
+                } else {
+                    ty = &ty[idx+1..];
+                }
+            }
+        } else {
+            if let Some(mut_idx) = ty[0..].find("mut") {
+                mutability.push(true);
+                ty = &ty[mut_idx+4..];
+            } else {
+                mutability.push(false);
+                if ty.contains("const") {
+                    ty = &ty[6..];
+                }
+            }
+        }
+    }
+    let c_ty = c_type(ty);
+    let mutability: Vec<&'static str> = mutability.iter().rev().map(|&m| if m { "*" } else { "const *" }).collect();
+    let result = format!("{} {}", c_ty, mutability.join(" "));
+    result
+}
+
+fn c_type(ty: &str) -> String {
+    match ty {
+        v if is_c_ptr(v) => { rust_ptr_to_c_ptr(v) }
+        "c_void" => "void".to_string(),
+        "boolean" => "boolean".to_string(),
+        "u8" => "unsigned char".to_string(),
+        "c_int" => "int".to_string(),
+        "f32" => "float".to_string(),
+        "c_ulong" => "unsigned long".to_string(),
+        "c_uint" => "unsigned int".to_string(),
+        "usize" => "uintptr_t".to_string(),
+        "JSAMPARRAY_MUT" => "JSAMPARRAY".to_string(),
+        "JSAMPIMAGE_MUT" => "JSAMPIMAGE".to_string(),
+        v => v.to_string(),
+    }
+}
+
+fn c_param(id: &str, ty: &str) -> String {
+    match ty {
+        v => format!("{} {}", c_type(ty), id),
+    }
+}
+
+fn cxx_header() -> String {
+r#"
+    #include <jinclude.h>
+    #include <jpeglib.h>
+    #include <jconfig.h>
+    #include <jconfigint.h>
+    #include <stdexcept>
+
+    struct WrapperError {
+        char* buf;
+        uintptr_t len;
+        uintptr_t cap;
+    };
+
+    struct mozjpeg_rust_wrapper_exception: std::runtime_error {
+        WrapperError error;
+    };
```

Comment:
This is the exception throw. It is just a std::runtime_exception with a Rust string payload.